### ISSUE TITLE
Question about the encoder for the Transformer

### ISSUE THREAD

**Issue by jlim13** [bentrevett/pytorch-seq2seq#116]
How to interpret the output of the encoder? For RNNs the output of the encoder is a single context vector that represents the entire input. This is of shape [batch x hidden_dimension_size]. The output of the transformer is of size [batch x sequence length x hidden_dimension_size]. The output of the encoder is a sequence of context vectors... Is there a way to represent this sequence of context vectors into one vector? How would you go from a sequence of context vectors from an encoder to a binary classification? 

Thanks!

**Comment by bentrevett** [comment]
This is referenced in the tutorial under the **Encoder** heading:

>Similar to the ConvSeq2Seq model, the Transformer's encoder does not attempt to compress the entire source sentence, $X = (x_1, ... ,x_n)$, into a single context vector, $z$. Instead it produces a sequence of context vectors, $Z = (z_1, ... , z_n)$. So, if our input sequence was 5 tokens long we would have $Z = (z_1, z_2, z_3, z_4, z_5)$. Why do we call this a sequence of context vectors and not a sequence of hidden states? A hidden state at time $t$ in an RNN has only seen tokens $x_t$ and all the tokens before it. However, each context vector here has seen all tokens at all positions within the input sequence.

As for using the Transformer's encoder for binary classification. I would following the [BERT](https://arxiv.org/pdf/1810.04805.pdf) paper (figure 4 on page 15). They simply feed the first (sequentially) context vector through a linear layer. So you would pass the sequence through the Transformer as normal and get the [batch, sequence length, hid dim] tensor, then index the first element in the sequence dimension and pass it through an `nn.Linear`, which would look something like:

```python
enc_src = self.encoder(src, src_mask)
out = self.fc(enc_src[:,0,:])
```

**Comment by bentrevett** [comment]
I'm not sure there is a set way of doing it that is "the best" way to do it.

If we look at the HuggingFace library, they just feed a "pooled" output through a linear layer: https://github.com/huggingface/transformers/blob/d541938c48f759522f81fa177aae49098e0e0149/src/transformers/modeling_bert.py#L1273

Their "pooled" output is actually just the first context vector, as done in the BERT paper: https://github.com/huggingface/transformers/blob/d541938c48f759522f81fa177aae49098e0e0149/src/transformers/modeling_bert.py#L666

However, even they themselves admit that it isn't the best method and that ideally your should actually pool or average across the hidden dimensions: https://github.com/huggingface/transformers/blob/d541938c48f759522f81fa177aae49098e0e0149/src/transformers/modeling_bert.py#L671
